### PR TITLE
feat(core): new topology and extensions manager definition

### PR DESCRIPTION
**Pull request by victorvassilev**

### CODE REVIEW

File: src/Centreon/Application/Webservice/TopologyWebservice.php [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+     */
+    public static function getName(): string
+    {
+        return 'centreon_topology_webservice';
+    }
+
+    /**
+     * @OA\Get(
+     *   path="/internal.php?object=centreon_topology_webservice&action=getTopologyByPage",
+     *   summary="Get topology object by page id",
```
Comment (vhr) [inline src/Centreon/Application/Webservice/TopologyWebservice.php]:
```suggestion
     *   description="Get topology object by page id",
```

File: features/TopologyAPI.feature [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+  In order to utilize topology params in React
+  As a logged in user
+  I need to ensure my API returns proper JSON for topology object
+
+  Background:
+    Given I am logged in a Centreon server
+    And I have a running instance of Centreon API
+
+  Scenario: Healthcheck
+    When I make a GET request to "/api/index.php?object=centreon_topology_webservice&action=getTopologyByPage&topology_page=1"
```
Comment (vhr) [inline features/TopologyAPI.feature]:
```suggestion
    When I make a GET request to "/api/index.php?object=centreon_topology&action=getTopologyByPage&topology_page=1"
```

File: src/Centreon/Application/Webservice/TopologyWebservice.php [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+{
+
+    /**
+     * Name of web service object
+     * 
+     * @return string
+     */
+    public static function getName(): string
+    {
+        return 'centreon_topology_webservice';
```
Comment (vhr) [inline src/Centreon/Application/Webservice/TopologyWebservice.php]:
```suggestion
        return 'centreon_topology';
```

File: src/Centreon/Application/Webservice/TopologyWebservice.php [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+    }
+
+    /**
+     * @OA\Get(
+     *   path="/internal.php?object=centreon_topology_webservice&action=getTopologyByPage",
+     *   summary="Get topology object by page id",
+     *   tags={"centreon_topology_webservice"},
```
Comment (vhr) [inline src/Centreon/Application/Webservice/TopologyWebservice.php]:
```suggestion
     *   tags={"centreon_topology"},
```

File: src/Centreon/Application/Webservice/TopologyWebservice.php [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+     *   @OA\Parameter(
+     *       in="query",
+     *       name="object",
+     *       @OA\Schema(
+     *          type="string",
+     *          enum={"centreon_topology_webservice"},
```
Comment (vhr) [inline src/Centreon/Application/Webservice/TopologyWebservice.php]:
```suggestion
     *          enum={"centreon_topology"},
```

Comment (vhr) [inline src/Centreon/Application/Webservice/TopologyWebservice.php]:
```suggestion
     *          enum={"centreon_topology"},
```

File: src/Centreon/Application/Webservice/TopologyWebservice.php [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+     *          default="centreon_topology_webservice"
```
Comment (vhr) [inline src/Centreon/Application/Webservice/TopologyWebservice.php]:
```suggestion
     *          default="centreon_topology"
```

File: src/Centreon/Application/Webservice/TopologyWebservice.php [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+     * @return string
+     */
+    public static function getName(): string
+    {
+        return 'centreon_topology_webservice';
+    }
+
+    /**
+     * @OA\Get(
+     *   path="/internal.php?object=centreon_topology_webservice&action=getTopologyByPage",
```
Comment (vhr) [inline src/Centreon/Application/Webservice/TopologyWebservice.php]:
```suggestion
     *   path="/internal.php?object=centreon_topology&action=getTopologyByPage",
```

File: src/CentreonModule/Infrastructure/Source/SourceDataInterface.php [hunk 8]
```diff
@@ -0,0 +1,42 @@
+<?php
+/*
+ * Copyright 2005-2019 Centreon
+ * Centreon is developped by : Julien Mathis and Romain Le Merlus under
```
Comment (sc979) [inline src/CentreonModule/Infrastructure/Source/SourceDataInterface.php]:
```suggestion
 * Centreon is developed by : Julien Mathis and Romain Le Merlus under
```

File: src/CentreonModule/Infrastructure/Source/SourceDataInterface.php [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+ * For more information : contact@centreon.com
+ *
+ *
+ */
+
+namespace CentreonModule\Infrastructure\Source;
+
+interface SourceDataInterface
+{
+
```
Comment (sc979) [inline src/CentreonModule/Infrastructure/Source/SourceDataInterface.php]: ?

File: src/CentreonModule/Infrastructure/Source/SourceInterface.php [hunk 10]
```diff
@@ -0,0 +1,54 @@
+<?php
+/*
+ * Copyright 2005-2019 Centreon
+ * Centreon is developped by : Julien Mathis and Romain Le Merlus under
```
Comment (sc979) [inline src/CentreonModule/Infrastructure/Source/SourceInterface.php]:
```suggestion
 * Centreon is developed by : Julien Mathis and Romain Le Merlus under
```

File: src/CentreonModule/Infrastructure/Source/SourceInterface.php [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+ * For more information : contact@centreon.com
+ *
+ *
+ */
+
+namespace CentreonModule\Infrastructure\Source;
+
+use CentreonModule\Infrastructure\Entity\Module;
+
+interface SourceInterface
```
Comment (sc979) [inline src/CentreonModule/Infrastructure/Source/SourceInterface.php]: document your code

File: src/CentreonModule/Infrastructure/Source/SourceInterface.php [hunk 12]
```diff
@@ -0,0 +1,54 @@
...
+{
+
+    public function getList(string $search = null, bool $installed = null, bool $updated = null);
```
Comment (sc979) [inline src/CentreonModule/Infrastructure/Source/SourceInterface.php]: WIP

File: src/CentreonModule/Infrastructure/Source/SourceInterface.php [hunk 13]
```diff
@@ -0,0 +1,54 @@
...
+
+    public function createEntityFromConfig(string $configFile): Module;
+
+    public function isEligible(
+        Module $entity,
+        string $search = null,
+        bool $installed = null,
+        bool $updated = null
+    ): bool;
+}
```
Comment (sc979) [inline src/CentreonModule/Infrastructure/Source/SourceInterface.php]: WIP

File: src/CentreonModule/Infrastructure/Source/WidgetSource.php [hunk 14]
```diff
@@ -0,0 +1,127 @@
+<?php
+/*
+ * Copyright 2005-2019 Centreon
+ * Centreon is developped by : Julien Mathis and Romain Le Merlus under
```
Comment (sc979) [inline src/CentreonModule/Infrastructure/Source/WidgetSource.php]:
```suggestion
 * Centreon is developed by : Julien Mathis and Romain Le Merlus under
```

File: src/CentreonModule/Infrastructure/Source/WidgetSource.php [hunk 15]
```diff
@@ -0,0 +1,127 @@
...
+ */
+
+namespace CentreonModule\Infrastructure\Source;
+
+use Psr\Container\ContainerInterface;
+use CentreonModule\Infrastructure\Entity\Module;
+use CentreonModule\Domain\Repository\WidgetModelsRepository;
+use CentreonModule\Infrastructure\Source\SourceAbstract;
+
+class WidgetSource extends SourceAbstract
```
Comment (sc979) [inline src/CentreonModule/Infrastructure/Source/WidgetSource.php]: document your class

File: src/CentreonModule/Infrastructure/Source/WidgetSource.php [hunk 16]
```diff
@@ -0,0 +1,127 @@
...
+     * @param \Psr\Container\ContainerInterface $services
+     */
+    public function __construct(ContainerInterface $services)
+    {
+        parent::__construct($services);
+
+        $this->info = $this->db
+            ->getRepository(WidgetModelsRepository::class)
+            ->getAllWidgetVsVersion()
+        ;
```
Comment (sc979) [inline src/CentreonModule/Infrastructure/Source/WidgetSource.php]: semi-colon -> check line 78

File: src/CentreonModule/Infrastructure/Source/WidgetSource.php [hunk 17]
```diff
@@ -0,0 +1,127 @@
...
+
+        return $result;
+    }
+
+    public function createEntityFromConfig(string $configFile): Module
+    {
+        $xml = simplexml_load_file($configFile);
+
+        $entity = new Module;
+        $entity->setId(basename(dirname($configFile)));
```
Comment (sc979) [inline src/CentreonModule/Infrastructure/Source/WidgetSource.php]: could $configFile be empty ? -> what happens if $files is empty ?

File: src/CentreonModule/ServiceProvider.php [hunk 18]
```diff
@@ -0,0 +1,78 @@
+<?php
+/*
+ * Copyright 2005-2019 Centreon
+ * Centreon is developped by : Julien Mathis and Romain Le Merlus under
```
Comment (sc979) [inline src/CentreonModule/ServiceProvider.php]:
```suggestion
 * Centreon is developed by : Julien Mathis and Romain Le Merlus under
```

**Pull request closed without merging**